In [60]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [45]:
#importing data
population=pd.read_csv("C:/Users/user/Downloads/Summary_Population_Distribution_By_District.csv")

In [46]:
#copying a new dataframe from the original so that we maintain the original copy
pop=population.copy()

In [47]:
pop.head(3)

_District rural/urban  Male  Female  Total  Households  Area_in_Sq_Km  \
0   NAIROBI WEST       Rural     0       0      0           0            0.0   
1   NAIROBI EAST       Rural     0       0      0           0            0.0   
2  NAIROBI NORTH       Rural     0       0      0           0            0.0   

   _Household_Density_  Population_Density  Gender_Index   County  \
0                  NaN                 NaN           NaN  NAIROBI   
1                  NaN                 NaN           NaN  NAIROBI   
2                  NaN                 NaN           NaN  NAIROBI   

                         Location  Location_1   Province  OBJECTID  
0   NAIROBI WEST, NAIROBI Kenya\n         NaN  36.799202         0  
1   NAIROBI EAST, NAIROBI Kenya\n         NaN  36.967484         1  
2  NAIROBI NORTH, NAIROBI Kenya\n         NaN  36.891126         2

In [48]:
#finding the number of rows and columns in the data
pop.shape

(316, 15)

In [49]:
#finding the types of data in different columns
pop.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 316 entries, 0 to 315
Data columns (total 15 columns):
_District              316 non-null object
rural/urban            316 non-null object
Male                   316 non-null int64
Female                 316 non-null int64
Total                  316 non-null int64
Households             316 non-null int64
Area_in_Sq_Km          316 non-null float64
_Household_Density_    290 non-null float64
Population_Density     290 non-null float64
Gender_Index           290 non-null float64
County                 316 non-null object
Location               316 non-null object
Location_1             0 non-null float64
Province               316 non-null float64
OBJECTID               316 non-null int64
dtypes: float64(6), int64(5), object(4)
memory usage: 37.1+ KB


In [50]:
# Total count of missing values in different columns
pop.isna().sum()

_District                0
rural/urban              0
Male                     0
Female                   0
Total                    0
Households               0
Area_in_Sq_Km            0
_Household_Density_     26
Population_Density      26
Gender_Index            26
County                   0
Location                 0
Location_1             316
Province                 0
OBJECTID                 0
dtype: int64

In [51]:
pop[['_Household_Density_','Population_Density','Gender_Index']].head()

_Household_Density_  Population_Density  Gender_Index
0                  NaN                 NaN           NaN
1                  NaN                 NaN           NaN
2                  NaN                 NaN           NaN
3                  NaN                 NaN           NaN
4                37.64         109861.0821      1.034624

In [52]:
pop.columns

Index(['_District', 'rural/urban', 'Male', 'Female', 'Total', 'Households',
       'Area_in_Sq_Km', '_Household_Density_', 'Population_Density',
       'Gender_Index', 'County', 'Location', 'Location_1', 'Province',
       'OBJECTID'],
      dtype='object')

### Data cleaning

In [53]:
#dropping the 'Location_1' column which has no values
pop.drop('Location_1',axis=1,inplace=True)

#finding the mean of House_density column and using the mean to fill the empty column
pop['_Household_Density_'].fillna(pop['_Household_Density_'].mean(),inplace=True)

#finding the mean of Population_density column and using the mean to fill the empty column
pop['Population_Density'].fillna(pop['Population_Density'].mean(),inplace=True)

#finding the mean of Gender_index column and using the mean to fill the empty column
pop['Gender_Index'].fillna(pop['Gender_Index'].mean(),inplace=True)



In [54]:
pop.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 316 entries, 0 to 315
Data columns (total 14 columns):
_District              316 non-null object
rural/urban            316 non-null object
Male                   316 non-null int64
Female                 316 non-null int64
Total                  316 non-null int64
Households             316 non-null int64
Area_in_Sq_Km          316 non-null float64
_Household_Density_    316 non-null float64
Population_Density     316 non-null float64
Gender_Index           316 non-null float64
County                 316 non-null object
Location               316 non-null object
Province               316 non-null float64
OBJECTID               316 non-null int64
dtypes: float64(5), int64(5), object(4)
memory usage: 34.6+ KB


### Descriptive analysis

In [56]:
pop.describe().T

count           mean            std         min  \
Male                 316.0   60735.626582   67904.042331    0.000000   
Female               316.0   61448.224684   67267.211785    0.000000   
Total                316.0  122183.851266  135017.724952    0.000000   
Households           316.0   27746.689873   37068.244631    0.000000   
Area_in_Sq_Km        316.0    1839.598608    4620.038177    0.000000   
_Household_Density_  316.0     601.943552     953.486229    0.400000   
Population_Density   316.0   67302.792308   64820.377111  486.702929   
Gender_Index         316.0       1.014161       0.075898    0.770335   
Province             316.0      36.583759       1.827359   34.020468   
OBJECTID             316.0     157.500000      91.365566    0.000000   

                              25%           50%            75%           max  
Male                 12418.250000  45400.000000   88108.250000  5.825540e+05  
Female               12040.250000  46362.500000   90530.750000  5.618620e+05  
Total                24396.500000  91507.500000  178853.500000  1.144416e+06  
Households            6043.250000  19699.500000   36117.500000  3.698660e+05  
Area_in_Sq_Km           32.182500    218.730000    1015.130000  3.924803e+04  
_Household_Density_     40.517500    202.545000     789.902500  9.720770e+03  
Population_Density   22831.982972  60329.002705   90614.262475  5.644318e+05  
Gender_Index             0.985581      1.016525       1.059230  1.428571e+00  
Province                34.902118     36.537310      37.625228  4.129295e+01  
OBJECTID                78.750000    157.500000     236.250000  3.150000e+02

In [57]:
#correlation of the whole data(numeric)
pop.corr()

Male    Female     Total  Households  Area_in_Sq_Km  \
Male                 1.000000  0.995459  0.998875    0.963354       0.139904   
Female               0.995459  1.000000  0.998853    0.959154       0.111252   
Total                0.998875  0.998853  1.000000    0.962357       0.125788   
Households           0.963354  0.959154  0.962357    1.000000       0.017582   
Area_in_Sq_Km        0.139904  0.111252  0.125788    0.017582       1.000000   
_Household_Density_  0.228824  0.199303  0.214376    0.367440      -0.239308   
Population_Density   0.958808  0.961719  0.961348    0.934537       0.079577   
Gender_Index        -0.028712  0.037656  0.004321    0.014071      -0.366865   
Province             0.000303 -0.032749 -0.016164   -0.025833       0.277585   
OBJECTID            -0.321564 -0.334824 -0.328536   -0.226969      -0.369777   

                     _Household_Density_  Population_Density  Gender_Index  \
Male                            0.228824            0.958808     -0.028712   
Female                          0.199303            0.961719      0.037656   
Total                           0.214376            0.961348      0.004321   
Households                      0.367440            0.934537      0.014071   
Area_in_Sq_Km                  -0.239308            0.079577     -0.366865   
_Household_Density_             1.000000            0.213907     -0.011672   
Population_Density              0.213907            1.000000      0.039251   
Gender_Index                   -0.011672            0.039251      1.000000   
Province                        0.010904           -0.045520     -0.465249   
OBJECTID                        0.390779           -0.322717      0.040782   

                     Province  OBJECTID  
Male                 0.000303 -0.321564  
Female              -0.032749 -0.334824  
Total               -0.016164 -0.328536  
Households          -0.025833 -0.226969  
Area_in_Sq_Km        0.277585 -0.369777  
_Household_Density_  0.010904  0.390779  
Population_Density  -0.045520 -0.322717  
Gender_Index        -0.465249  0.040782  
Province             1.000000 -0.308864  
OBJECTID            -0.308864  1.000000

In [58]:
#variable that have very high correlation
pop[['Male','Female','Households','Population_Density']].corr()

Male    Female  Households  Population_Density
Male                1.000000  0.995459    0.963354            0.958808
Female              0.995459  1.000000    0.959154            0.961719
Households          0.963354  0.959154    1.000000            0.934537
Population_Density  0.958808  0.961719    0.934537            1.000000

In [59]:
### The next step is to divide the data into attributes and labels 
X=pop[['Male','Female','Households']]
y=pop['Population_Density']

### Implementing multiple linear regeression


In [62]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

### Training the algorithm

In [63]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [65]:
#the three coefficients of our multiple linear regression
coeff_df = pd.DataFrame(regressor.coef_, X.columns, columns=['Coefficient'])
coeff_df

Coefficient
Male           0.132255
Female         0.634572
Households     0.260681

### Making prediction

In [66]:
y_pred = regressor.predict(X_test)

In [67]:
y_pred

array([135071.45630984,  19257.88276763,  83567.71804108,  13054.26611261,
        13054.26611261, 157766.61745933,  75489.08333514,  75389.76658382,
        33203.51618812, 111630.63296634,  13690.70561878,  54052.93513511,
        71425.57235   ,  92313.09201763,  18545.90067504,  67069.73622153,
        32306.40107534, 121324.86666901, 140473.8556512 ,  62900.35782187,
       165463.2064419 , 543058.56551315,  45030.50564667,  41223.56841205,
        24962.59888805,  42996.54395615,  62301.56077308,  13054.26611261,
       112157.74676766,  27057.41175308, 124097.76149   ,  13054.26611261,
        49976.88300723, 115664.49291532,  13428.33259223,  13667.71158371,
        73287.54924167,  55086.62501741,  22115.69534651,  21335.11359907,
        40832.07030867,  43556.84294345,  19590.04842632, 135964.8653636 ,
        16634.00287828, 143854.89482983,  62350.51512467,  55344.48926395,
        13054.26611261, 106457.02093419,  62450.31083526, 188304.17978119,
        17796.72055748,  

In [68]:
#being an numpy array we convert the array into list appending the values into a new list called y_pred_final and compare the predicted values against the test values

y_pred_final=[]
for i in y_pred:
    y_pred_final.append(int(i))
    
y_pred_final

[135071,
 19257,
 83567,
 13054,
 13054,
 157766,
 75489,
 75389,
 33203,
 111630,
 13690,
 54052,
 71425,
 92313,
 18545,
 67069,
 32306,
 121324,
 140473,
 62900,
 165463,
 543058,
 45030,
 41223,
 24962,
 42996,
 62301,
 13054,
 112157,
 27057,
 124097,
 13054,
 49976,
 115664,
 13428,
 13667,
 73287,
 55086,
 22115,
 21335,
 40832,
 43556,
 19590,
 135964,
 16634,
 143854,
 62350,
 55344,
 13054,
 106457,
 62450,
 188304,
 17796,
 24313,
 35102,
 217477,
 80104,
 13054,
 51663,
 61062,
 117764,
 13816,
 37212,
 125433,
 134199,
 33662,
 61990,
 26886,
 26482,
 68015,
 24847,
 211482,
 83104,
 33835,
 90951,
 91362,
 121483,
 22057,
 78072,
 40072,
 64435,
 13776,
 248201,
 28482,
 22275,
 75513,
 101514,
 61199,
 52901,
 108850,
 142884,
 247046,
 21228,
 31868,
 176842]

In [77]:
y_real=list(y_test)

In [75]:
# convert the outcome into a dictionary/dataframe for ease of comparison
data={'Predicted':y_pred_final,'Test_values':y_real}
data_df=pd.DataFrame(data)
data_df


Predicted    Test_values
0      135071  138775.318000
1       19257    7425.700162
2       83567   81417.279340
3       13054   67302.792308
4       13054   67302.792308
5      157766  168157.399700
6       75489   72213.970330
7       75389   71038.560840
8       33203   23645.071200
9      111630  114010.901600
10      13690     720.921505
11      54052   48011.306380
12      71425   65433.744510
13      92313   92386.101630
14      18545    6175.901542
15      67069   61689.525070
16      32306   21127.155780
17     121324  126013.030500
18     140473  149016.203400
19      62900   56751.145630
20     165463  182919.028800
21     543058  564431.826600
22      45030   35778.827610
23      41223   32923.160730
24      24962   14276.670430
25      42996   34042.789010
26      62301   55973.113370
27      13054   67302.792308
28     112157  114066.710600
29      27057   15934.235190
..        ...            ...
65      33662   22836.158150
66      61990   54698.312200
67      26886   15822.713630
68      26482   15187.219160
69      68015   63315.473090
70      24847   14037.441780
71     211482  228665.817900
72      83104   81153.705850
73      33835   23933.314290
74      90951   88172.622500
75      91362   91322.218930
76     121483  122884.765100
77      22057   10566.291570
78      78072   76214.825550
79      40072   31244.935860
80      64435   57357.715190
81      13776    1153.522843
82     248201  271574.346600
83      28482   18095.349770
84      22275   10848.932450
85      75513   72792.951760
86     101514  103154.709800
87      61199   55721.448570
88      52901   46555.398130
89     108850  108056.847500
90     142884  150340.119200
91     247046  257268.241400
92      21228    9840.321045
93      31868   21680.141120
94     176842  191378.588300

[95 rows x 2 columns]

### Model evaluation

In [78]:
from sklearn import metrics
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

Mean Absolute Error: 10838.822230998594
Mean Squared Error: 264803223.29735506
Root Mean Squared Error: 16272.775525316973


In [79]:
pop['Population_Density'].mean()

67302.79230761167

In [81]:
0.1*67302.79230761167


6730.279230761167

### You can see that the value of root mean squared error is 16272.78, which is greater than 10% of the mean value of the population density in all districts.This means that our algorithm was not very accurate but can still make reasonably good predictions.